File: vipaq/src/Binacle.ViPaq/HeaderNotation.cs
namespace Binacle.ViPaq;

// The text form of a `Header`, so a test vector can name the exact header its bytes were produced under
// (PROTOCOL.md §6.1: byte comparison only means something once the header is pinned).
//
// The grammar, six underscore-delimited tokens in wire order:
//
//   v{N}_{raw|comp}_{row|col}_{binW}_{itemDimW}_{itemCoordW}
//
//   v1_raw_row_8_8_8      uncompressed, row-major, all 8-bit (an empty or small pack)
//   v1_comp_col_16_8_16   compressed, columnar, 16-bit bin dims, 8-bit item dims, 16-bit item coords (Bischoff)
//
// Version leads and is mandatory: that is what makes positional parsing forward-safe, since the header cannot
// gain a field without a version bump. The compressed flag is a bare bit - never `deflate` or `gzip`, because
// the wire carries no codec field (PROTOCOL.md §6). Widths are only `8` or `16`; reserved codes 2 and 3 never
// reach the wire, so they have no notation.
//
// A malformed notation string is a bad label, not a bad blob, so `Parse` throws `FormatException`.
// `ViPaqFormatException` stays reserved for wire bytes.
internal static class HeaderNotation
{
	private const char Separator = '_';
	private const int TokenCount = 6;

	public static string Format(Header header)
	{
		return string.Join(
			Separator,
			FormatVersion(header.Version),
			header.Compressed ? "comp" : "raw",
			FormatLayout(header.Layout),
			FormatWidth(header.BinDimensionsWidth),
			FormatWidth(header.ItemDimensionsWidth),
			FormatWidth(header.ItemCoordinatesWidth)
			);
	}

	public static Header Parse(string notation)
	{
		ArgumentNullException.ThrowIfNull(notation);

		var tokens = notation.Split(Separator);
		if (tokens.Length != TokenCount)
		{
			throw new FormatException(
				$"Header notation must have {TokenCount} tokens separated by '{Separator}', got {tokens.Length}: '{notation}'"
				);
		}

		return new Header
		{
			Version = ParseVersion(tokens[0]),
			Compressed = ParseCompressed(tokens[1]),
			Layout = ParseLayout(tokens[2]),
			BinDimensionsWidth = ParseWidth(tokens[3], "bin dimensions"),
			ItemDimensionsWidth = ParseWidth(tokens[4], "item dimensions"),
			ItemCoordinatesWidth = ParseWidth(tokens[5], "item coordinates")
		};
	}

	// Only Version1 has a spelling - the one version written. Reserved codes never reach the wire.
	private static string FormatVersion(Version version)
	{
		return version switch
		{
			Version.Version1 => "v1",
			_ => throw new ArgumentOutOfRangeException(
				nameof(version),
				version,
				$"Only {Version.Version1} has a notation"
				)
		};
	}

	private static Version ParseVersion(string token)
	{
		return token switch
		{
			"v1" => Version.Version1,
			_ => throw new FormatException($"Unknown version token '{token}', this implementation reads 'v1'")
		};
	}

	private static bool ParseCompressed(string token)
	{
		return token switch
		{
			"comp" => true,
			"raw" => false,
			_ => throw new FormatException($"Unknown compression token '{token}', expected 'raw' or 'comp'")
		};
	}

	private static string FormatLayout(Layout layout)
	{
		return layout switch
		{
			Layout.RowMajor => "row",
			Layout.Columnar => "col",
			_ => throw new ArgumentOutOfRangeException(nameof(layout), layout, "Unknown layout")
		};
	}

	private static Layout ParseLayout(string token)
	{
		return token switch
		{
			"row" => Layout.RowMajor,
			"col" => Layout.Columnar,
			_ => throw new FormatException($"Unknown layout token '{token}', expected 'row' or 'col'")
		};
	}

	private static string FormatWidth(Width width)
	{
		return width switch
		{
			Width.Eight => "8",
			Width.Sixteen => "16",
			_ => throw new ArgumentOutOfRangeException(nameof(width), width, "Reserved width has no notation")
		};
	}

	private static Width ParseWidth(string token, string section)
	{
		return token switch
		{
			"8" => Width.Eight,
			"16" => Width.Sixteen,
			_ => throw new FormatException($"Unknown width token '{token}' for {section}, expected '8' or '16'")
		};
	}
}

File: vipaq/packages/binacle-vipaq/src/headerNotation.ts
import {Header, Layout, Version, Width} from "./models";

// Ports C#: HeaderNotation. Text form of a Header, used by the test vectors and the interop generator so a
// vector can name the exact header its bytes were produced under. Wire-specific (it names Header/Width/Layout/
// Version), so it stays in the vipaq mirror; the geometry notation lives in binacle-compact-notation.
//
// Grammar (six underscore-delimited tokens, in wire order):
//
//   v{N}_{raw|comp}_{row|col}_{binW}_{itemDimW}_{itemCoordW}
//
//   v1_raw_row_8_8_8      uncompressed, row-major, all 8-bit
//   v1_raw_col_16_16_16   uncompressed, columnar, all 16-bit
//
// Version leads and is mandatory (positional parsing stays forward-safe). The compressed flag is a bare bit
// (`comp`, never a codec name) — the header carries no codec. Widths are only `8` or `16`; reserved codes never
// reach the wire, so they never reach the notation. A bad label is not a bad blob, so a bad token throws a plain
// Error (mirrors C# throwing FormatException, not ViPaqFormatException).

const separator = "_";
const tokenCount = 6;

export function parseHeader(notation: string): Header {
	const tokens = notation.split(separator);
	if (tokens.length !== tokenCount) {
		throw new Error(
			`Header notation must have ${tokenCount} tokens separated by '${separator}', got ${tokens.length}: '${notation}'`,
		);
	}
	return new Header(
		parseVersion(tokens[0]),
		parseCompressed(tokens[1]),
		parseLayout(tokens[2]),
		parseWidth(tokens[3], "bin dimensions"),
		parseWidth(tokens[4], "item dimensions"),
		parseWidth(tokens[5], "item coordinates"),
	);
}

export function formatHeader(header: Header): string {
	return [
		formatVersion(header.version),
		header.compressed ? "comp" : "raw",
		formatLayout(header.layout),
		formatWidth(header.binDimensionsWidth),
		formatWidth(header.itemDimensionsWidth),
		formatWidth(header.itemCoordinatesWidth),
	].join(separator);
}

function parseVersion(token: string): Version {
	if (token === "v1") return Version.Version1;
	throw new Error(`Unknown version token '${token}', this implementation reads 'v1'`);
}

function formatVersion(version: Version): string {
	if (version === Version.Version1) return "v1";
	throw new Error(`Only version ${Version.Version1} has a notation`);
}

function parseCompressed(token: string): boolean {
	if (token === "comp") return true;
	if (token === "raw") return false;
	throw new Error(`Unknown compression token '${token}', expected 'raw' or 'comp'`);
}

function parseLayout(token: string): Layout {
	if (token === "row") return Layout.RowMajor;
	if (token === "col") return Layout.Columnar;
	throw new Error(`Unknown layout token '${token}', expected 'row' or 'col'`);
}

function formatLayout(layout: Layout): string {
	if (layout === Layout.RowMajor) return "row";
	if (layout === Layout.Columnar) return "col";
	throw new Error(`Unknown layout ${layout}`);
}

function parseWidth(token: string, section: string): Width {
	if (token === "8") return Width.Eight;
	if (token === "16") return Width.Sixteen;
	throw new Error(`Unknown width token '${token}' for ${section}, expected '8' or '16'`);
}

function formatWidth(width: Width): string {
	if (width === Width.Eight) return "8";
	if (width === Width.Sixteen) return "16";
	throw new Error(`Reserved width ${width} has no notation`);
}
